pootle/apps/pootle_fs/finder.py

```diff
@@ -0,0 +1,182 @@
+
+
+class TranslationFileFinder(object):
+
+    path_mapping = PATH_MAPPING
+
+    def __init__(self, translation_path, path_filters=None, ext=None):
+        self.translation_path = translation_path
+        self.ext = ext or DEFAULT_EXTS
+        self.path_filters = path_filters
+        TranslationPathValidator(translation_path).validate()
+
+    @cached_property
+    def regex(self):
+        return re.compile(self._parse_path())
+
+    @cached_property
+    def file_root(self):
+        file_root = self.translation_path.split("<")[0]
+        if not file_root.endswith("/"):
+            file_root = "/".join(file_root.split("/")[:-1])
+        return file_root.rstrip("/")
+
+    def match_filepath(self, file_path):
+        filename = os.path.basename(file_path)
+        match = self.match(file_path)
+        if not match:
+            return
+        matched = match.groupdict()
+        matched["dir_path"] = (
+            matched.get("dir_path", "").strip("/"))
+        if not matched.get("filename"):
+            matched["filename"] = os.path.splitext(
+                os.path.basename(filename))[0]
+        return file_path, matched
+
+    def walk(self):
+        # print("Walking the FS: %s" % self.file_root)
+        for root, dirs, files in scandir.walk(self.file_root):
+            for filename in files:
+                yield os.path.join(root, filename)
+
+    def find(self):
+        for filepath in self.walk():
+            match = self.match_filepath(filepath)
+            if match:
+                yield match
+
+    def match(self, file_path):
+        return self.regex.match(file_path)
+
+    @lru_cache(maxsize=None)
+    def reverse_match(self, language_code, filename=None, ext=None, dir_path=None):
+        if not filename:
+            filename = language_code
+        if ext is None:
+            ext = self.ext[0]
+        if ext not in self.ext:
+            raise ValueError("ext must be in the list of possible extensions")
+        ext = ext.strip(".")
+        path = (
+            "%s.%s"
+            % (os.path.splitext(self.translation_path)[0], ext))
+        if dir_path and "<dir_path>" not in path:
+            return
+        path = (path.replace("<language_code>", language_code)
+                    .replace("<filename>", filename))
+        if "<dir_path>" in path:
+            if dir_path and dir_path.strip("/"):
+                path = path.replace(
+                    "<dir_path>", "/%s/" % dir_path.strip("/"))
+            else:
+                path = path.replace("<dir_path>", "")
+        local_path = path.replace(self.file_root, "")
+        if "//" in local_path:
+            path = os.path.join(
+                self.file_root,
+                local_path.replace("//", "/").lstrip("/"))
+        return path
+
+    def _ext_re(self):
+        return (
+            r".(?P<ext>(%s))"
+            % "|".join(
+                ("%s$" % x)
+                for x in set(self.ext)))
```

Comment:
Current code is fine. I am just not used to generator expressions. My apologies.